### PR TITLE
Compile fix for huge dependency graph

### LINKED ISSUE #3404: huge webpack-stats produce a JavaScript heap out of memory
We are working on a Very large project with several dependencies, when building the project we get a  `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`.

- [ x] I have searched the [issues](https://github.com/zeit/next.js/issues) of this repository and believe that this is not a duplicate.

## Expected Behavior

When running `yarn build` I should get a working production bundle

## Current Behavior

When running `yarn build` I get `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`.

After several hours of debugging I found out that the origin of this problem comes from https://github.com/zeit/next.js/blob/canary/server/build/index.js#L72
more exactly the `JSON.stringify` function, given that our stats file is 260MB, node can't handle it.

I was able to patch it using streams and make it work ok, I'll make a PR for this tomorrow.

## Your Environment

| Tech    | Version |
|---------|---------|
| next    |    4.1.4     |
| node    |    8.9.1     |
| OS      |     Ubuntu 16.04    |
| browser |    N/A     |

### PR DESCRIPTION
**Pull request by batusai513**
This PR solves an issue when trying to compile projects with many dependencies, right now, when next tries to write the huge dependency object into `webpack-stats.json`, the build trows with `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`, so i moved the JSON.stringify to use streams and write that stream using node fs module.

This PR fixes https://github.com/zeit/next.js/issues/3404 and https://github.com/zeit/next.js/issues/3391

### CODE REVIEW

File: server/build/index.js [hunk 1]
```diff
@@ -1,11 +1,14 @@
 import { tmpdir } from 'os'
 import { join } from 'path'
 import fs from 'mz/fs'
+import fs2 from 'fs'
```
Comment (tusbar) [inline server/build/index.js]: Maybe, `import { createWriteStream, createWriteStream } from 'fs'`, and move up along with the other native modules.

Comment (batusai513) [reply server/build/index.js]: Sounds reasonable!

**Comment by arunoda** [comment]
I think we should remove saving this at all.
It's a dev feature and not require all the times.

If needed, we can use [this technique](https://github.com/zeit/next.js/tree/canary/examples/with-webpack-bundle-analyzer) or similar.

**Comment by arunoda** [comment]
Could you simply writing this file at all?

**Comment by arunoda** [comment]
@timneutkens I think ^^^ is a good way. Isn't it.

**Review by arunoda — CHANGES_REQUESTED** [review]
Could you simply writing this file at all?
@timneutkens I think ^^^ is a good way. Isn't it?

**Comment by batusai513** [comment]
@arunoda I'm afraid you've lost me there, do you want me to remove completely the `webpack-stats.json` creation?

**Comment by arunoda** [comment]
@batusai513 yes. It's not needed for production stage.
It takes time and memory.

**Comment by batusai513** [comment]
@arunoda I've updated the PR

**Comment by timneutkens** [comment]
@batusai513 thank you very much, i've actually already removed it and did a new release (4.2.1) 👍

**Comment by batusai513** [comment]
@arunoda the change it's already on canary, thanks @timneutkens

**Pull request closed without merging**